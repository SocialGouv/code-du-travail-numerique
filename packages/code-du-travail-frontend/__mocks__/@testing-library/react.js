/* eslint-disable import/export */
import { theme } from "@socialgouv/cdtn-ui";
import { render } from "@testing-library/react";
import PropTypes from "prop-types";
import React from "react";
import { ThemeProvider } from "styled-components";

const AllTheProviders = ({ children }) => {
  return (
    <ThemeProvider theme={theme.colors}>
      <>{children}</>
    </ThemeProvider>
  );
};

AllTheProviders.propTypes = {
  children: PropTypes.node.isRequired,
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
