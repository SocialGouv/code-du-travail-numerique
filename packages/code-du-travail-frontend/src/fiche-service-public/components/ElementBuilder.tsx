import React from "react";
// import { useUIDSeed } from "react-uid";
import { getText, ignoreParagraph } from "../utils"; // import { LienExterne, LienExterneCommente } from "./LienExterne.js";
import { fr } from "@codegouvfr/react-dsfr";
import { LienExterne, LienExterneCommente } from "./LienExterne";
import List from "./List";
import Accordion from "./Accordion";
import Title from "./Title";
import Avertissement from "./Avertissement"; // Beware, this one is recursive

export type FicheSPDataText = { type: "text"; text: string };
export type FicheSPDataElement = {
  type: "element";
  children: FicheSPData[];
  name: string;
};
export type FicheSPData = {
  name?: string;
  attributes?: Record<string, string>;
} & (FicheSPDataElement | FicheSPDataText);

// Beware, this one is recursive
export const ElementBuilder = ({
  data,
  headingLevel = 0,
}: {
  data: FicheSPData | FicheSPData[];
  headingLevel?: number;
}) => {
  // const seedId = useUIDSeed();
  // in cases where the parent's "$"/children is undefined while it should not
  // e.g. in a "Texte" element. It occurs sometimes.
  if (!data) return <></>;
  // In case we get children

  if (Array.isArray(data)) {
    return data.map((child) => (
      <ElementBuilder key={"abc"} data={child} headingLevel={headingLevel} />
    ));
  }
  if (data.type === "text") {
    return data.text;
  }

  switch (data.name) {
    case "BlocCas":
      if (data.attributes?.affichage === "onglet") {
        return "Tabulator"; // <Tabulator data={data} headingLevel={headingLevel} />;
      } else {
        return <Accordion data={data} headingLevel={headingLevel} />;
      }
    case "Image":
      return "ImageElement"; //<ImageElement data={data} />;
    case "Introduction":
      if (ignoreParagraph(data)) {
        return (
          <p>
            <ElementBuilder data={ignoreParagraph(data)} />
          </p>
        );
      }
      break;
    case "LienExterne":
      return <LienExterne data={data} />;
    case "LienExterneCommente":
      return <LienExterneCommente data={data} />;
    case "Liste":
      return <List data={data} headingLevel={headingLevel} />;
    case "ListeSituations":
      return "ListeSituations"; // <Tabulator data={data} headingLevel={headingLevel} />;
    case "OuSAdresser":
      return "OuSAdresser"; // <OuSAdresser data={data} headingLevel={headingLevel} />;
    case "ServiceEnLigne":
    case "PourEnSavoirPlus":
      return "ServiceEnLigne"; // <ServiceEnLigne data={data} />;
    case "Tableau":
      return "Table"; // <Table data={data} headingLevel={headingLevel} />;
    case "Texte":
      if (data.children.find((child) => child.name === "Chapitre")) {
        return <Accordion data={data} headingLevel={headingLevel} />;
      }
      return (
        <ElementBuilder data={data.children} headingLevel={headingLevel} />
      );
    case "Titre":
      return <Title level={headingLevel}>{getText(data)}</Title>;
    // "Simple" elements, we can immediately parse their children
    case "Avertissement":
      return <Avertissement data={data} headingLevel={headingLevel} />;
    case "ANoter":
    case "ASavoir":
    case "Attention":
    case "Rappel":
      return (
        <div className={fr.cx("fr-callout", "fr-p-4w")}>
          <ElementBuilder data={data.children} headingLevel={headingLevel} />
        </div>
      );
    case "Chapitre":
    case "Description":
    case "FragmentConditionne":
    case "SousChapitre":
      return (
        <ElementBuilder data={data.children} headingLevel={headingLevel} />
      );
    case "Expression":
      return (
        <i>
          <ElementBuilder data={data.children} headingLevel={headingLevel} />
        </i>
      );
    case "MiseEnEvidence":
    case "Valeur":
      return (
        <strong>
          <ElementBuilder data={data.children} headingLevel={headingLevel} />
        </strong>
      );
    case "Paragraphe":
      return (
        <p>
          <ElementBuilder data={data.children} headingLevel={headingLevel} />
        </p>
      );
    // // TODO : check if exists These ones are still to be defined

    case "Exposant":
      return (
        <sup>
          <ElementBuilder data={data.children} headingLevel={headingLevel} />
        </sup>
      );
    case "LienIntra":
    case "LienInterne":
      return (
        <ElementBuilder data={data.children} headingLevel={headingLevel} />
      );
    // Otherwise we simply ignore the element
    default:
      return null;
  }
};
