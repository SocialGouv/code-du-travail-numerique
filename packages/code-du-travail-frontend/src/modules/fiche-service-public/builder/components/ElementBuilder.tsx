"use client";
import React from "react";
import { useUIDSeed } from "react-uid";
import { getText, ignoreParagraph } from "../utils";
import { LienExterne, LienExterneCommente } from "./LienExterne";
import List from "./List";

import Title from "./Title";
import Avertissement from "./Avertissement";
import Table from "./Table";
import ANoter from "./ANoter";
import ServiceEnLigne from "./ServiceEnLigne";
import OuSAdresser from "./OuSAdresser";
import Tabulator from "./Tabulator";
import { ImageComponent as ImageElement } from "./ImageComponent";
import {
  FicheSPData,
  FicheSPDataTextWithChapitre,
  isFicheSPDataChapitre,
} from "../type";
import AccordionWrapper from "./Accordion";
import SectionWithTitle from "./SectionWithTitle";
import TitreFlottant from "./TitreFlottant";
import Exemple from "./Exemple";
import Video from "./Video";

export const ElementBuilder = ({
  data,
  headingLevel = 0,
}: {
  data: FicheSPData | FicheSPData[];
  headingLevel?: number;
}) => {
  const seedId = useUIDSeed();
  // in cases where the parent's "$"/children is undefined while it should not
  // e.g. in a "Texte" element.
  if (!data) return <></>;

  if (Array.isArray(data)) {
    return data.map((child) => (
      <ElementBuilder
        key={seedId(child)}
        data={child}
        headingLevel={headingLevel}
      />
    ));
  }
  if (data.type === "text") {
    return data.text;
  }

  switch (data.name) {
    case "BlocCas":
      if (data.attributes.affichage === "onglet") {
        return <Tabulator data={data} headingLevel={headingLevel} />;
      }
      return (
        <ElementBuilder data={data.children} headingLevel={headingLevel} />
      );

    case "Image":
      return <ImageElement data={data} headingLevel={headingLevel} />;
    case "Introduction":
      if (ignoreParagraph(data)) {
        return (
          <p>
            <ElementBuilder data={ignoreParagraph(data)} />
          </p>
        );
      }
      break;
    case "LienExterne":
      return <LienExterne data={data} />;
    case "LienExterneCommente":
      return <LienExterneCommente data={data} />;
    case "Liste":
      return <List data={data} headingLevel={headingLevel} />;
    case "ListeSituations":
      return <Tabulator data={data} headingLevel={headingLevel} />;
    case "OuSAdresser":
      return <OuSAdresser data={data} headingLevel={headingLevel} />;
    case "ServiceEnLigne":
    case "PourEnSavoirPlus":
      return <ServiceEnLigne data={data} />;
    case "Tableau":
      return <Table data={data} headingLevel={headingLevel} />;
    case "Titre":
      return <Title level={headingLevel}>{getText(data)}</Title>;
    case "TitreFlottant":
      return <TitreFlottant level={headingLevel} data={data} />;
    case "Avertissement":
    case "Attention":
      return <Avertissement data={data} headingLevel={headingLevel} />;
    case "ANoter":
    case "ASavoir":
    case "Rappel":
      return <ANoter data={data} headingLevel={headingLevel} />;
    case "Exemple":
      return <Exemple data={data} headingLevel={headingLevel} />;
    case "Cas":
      return <SectionWithTitle data={data} headingLevel={headingLevel} />;
    case "Video":
      return <Video data={data} />;
    case "Texte":
      if (data.children.find((child) => isFicheSPDataChapitre(child))) {
        return (
          <AccordionWrapper
            data={data as FicheSPDataTextWithChapitre}
            headingLevel={headingLevel}
          />
        );
      }
      return (
        <ElementBuilder data={data.children} headingLevel={headingLevel} />
      );
    case "Description":
    case "FragmentConditionne":
    case "SousChapitre":
    case "LienIntra":
    case "LienInterne":
    case "Complement":
      return (
        <ElementBuilder data={data.children} headingLevel={headingLevel} />
      );
    case "Expression":
      return (
        <i>
          <ElementBuilder data={data.children} headingLevel={headingLevel} />
        </i>
      );
    case "MiseEnEvidence":
    case "Valeur":
      return (
        <strong>
          <ElementBuilder data={data.children} headingLevel={headingLevel} />
        </strong>
      );
    case "Paragraphe":
      return (
        <p>
          <ElementBuilder data={data.children} headingLevel={headingLevel} />
        </p>
      );
    case "Exposant":
      return (
        <sup>
          <ElementBuilder data={data.children} headingLevel={headingLevel} />
        </sup>
      );
    // Otherwise we simply ignore the element
    default:
      return null;
  }
};
