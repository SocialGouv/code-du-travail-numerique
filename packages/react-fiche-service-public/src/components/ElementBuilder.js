import { theme, Wrapper } from "@socialgouv/cdtn-ui";
import PropTypes from "prop-types";
import React from "react";
import { useUIDSeed } from "react-uid";
import styled from "styled-components";

import { getText, ignoreParagraph } from "../utils.js";
import Accordion from "./Accordion.js";
import { LienExterne, LienExterneCommente } from "./LienExterne.js";
import List from "./List.js";
import OuSAdresser from "./OuSAdresser.js";
import ServiceEnLigne from "./ServiceEnLigne.js";
import Table from "./Table.js";
import Tabulator from "./Tabulator.js";
import Title from "./Title.js";

const { spacings } = theme;

// Beware, this one is recursive
// parentAttributes only purpose for now is to pass a date to the title element
// from the parent level (Avertissement element), we might remove it later
export function ElementBuilder({ data, headingLevel, parentAttributes }) {
  const seedId = useUIDSeed();
  // in cases where the parent's "$"/children is undefined while it should not
  // e.g. in a "Texte" element. It occurs sometimes.
  if (!data) return null;
  // In case we get children
  if (Array.isArray(data)) {
    return data.map((child) => (
      <ElementBuilder
        key={seedId(child)}
        data={child}
        headingLevel={headingLevel}
        parentAttributes={parentAttributes}
      />
    ));
  }
  if (data.type === "text") {
    return data.text;
  }
  switch (data.name) {
    // Complex elements, we don't immediately parse their children
    case "BlocCas":
      if (data.attributes.affichage === "onglet") {
        return <Tabulator data={data} headingLevel={headingLevel} />;
      } else {
        return <Accordion data={data} headingLevel={headingLevel} />;
      }
    case "Introduction":
      if (ignoreParagraph(data)) {
        return (
          <Introduction>
            <ElementBuilder data={ignoreParagraph(data)} />
          </Introduction>
        );
      }
      break;
    case "LienExterne":
      return <LienExterne data={data} />;
    case "LienExterneCommente":
      return <LienExterneCommente data={data} />;
    case "Liste":
      return <List data={data} headingLevel={headingLevel} />;
    case "ListeSituations":
      return <Tabulator data={data} headingLevel={headingLevel} />;
    case "OuSAdresser":
      return <OuSAdresser data={data} headingLevel={headingLevel} />;
    case "ServiceEnLigne":
    case "PourEnSavoirPlus":
      return <ServiceEnLigne data={data} />;
    case "Tableau":
      return <Table data={data} headingLevel={headingLevel} />;
    case "Texte":
      if (data.children.find((child) => child.name === "Chapitre")) {
        return <Accordion data={data} headingLevel={headingLevel} />;
      }
      return (
        <ElementBuilder data={data.children} headingLevel={headingLevel} />
      );
    case "Titre":
      return (
        <Title level={headingLevel} date={parentAttributes.date}>
          {getText(data)}
        </Title>
      );
    // "Simple" elements, we can immediately parse their children
    case "ANoter":
    case "ASavoir":
    case "Attention":
    case "Avertissement":
    case "Rappel":
      return (
        <ANoter variant="dark">
          <ElementBuilder
            data={data.children}
            headingLevel={headingLevel + 1}
            parentAttributes={data.attributes}
          />
        </ANoter>
      );
    case "Chapitre":
    case "SousChapitre":
      return (
        <ElementBuilder data={data.children} headingLevel={headingLevel} />
      );
    case "Expression":
      return (
        <i>
          <ElementBuilder data={data.children} headingLevel={headingLevel} />
        </i>
      );
    case "MiseEnEvidence":
    case "Valeur":
      return (
        <strong>
          <ElementBuilder data={data.children} headingLevel={headingLevel} />
        </strong>
      );
    case "Paragraphe":
      return (
        <p>
          <ElementBuilder data={data.children} headingLevel={headingLevel} />
        </p>
      );
    case "Exposant":
      return (
        <sup>
          <ElementBuilder data={data.children} headingLevel={headingLevel} />
        </sup>
      );
    // These ones are still to be defined
    case "LienIntra":
    case "LienInterne":
      return (
        <ElementBuilder data={data.children} headingLevel={headingLevel} />
      );
    // Otherwise we simply ignore the element
    default:
      return null;
  }
}

ElementBuilder.propTypes = {
  data: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  headingLevel: PropTypes.number,
  parentAttributes: PropTypes.object,
};

ElementBuilder.defaultProps = {
  headingLevel: 0,
  parentAttributes: {},
};

const ANoter = styled(Wrapper)`
  margin-bottom: ${spacings.base};
`;

const Introduction = styled.p`
  margin-bottom: ${spacings.large};
`;
