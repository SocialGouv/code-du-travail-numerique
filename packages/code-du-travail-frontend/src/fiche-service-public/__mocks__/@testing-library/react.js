import React from "react";
import PropTypes from "prop-types";
import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import { theme } from "@socialgouv/cdtn-ui";

const AllTheProviders = ({ children }) => {
  return (
    <ThemeProvider theme={theme.colors}>
      <>{children}</>
    </ThemeProvider>
  );
};

AllTheProviders.propTypes = {
  children: PropTypes.node.isRequired,
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
