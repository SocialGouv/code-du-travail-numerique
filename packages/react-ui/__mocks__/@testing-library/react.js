import { render } from "@testing-library/react";
import PropTypes from "prop-types";
import React from "react";
import { ThemeProvider } from "styled-components";

import { colors } from "../../src/theme";

const AllTheProviders = ({ children }) => {
  return (
    <ThemeProvider theme={colors}>
      <>{children}</>
    </ThemeProvider>
  );
};

AllTheProviders.propTypes = {
  children: PropTypes.node.isRequired,
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
