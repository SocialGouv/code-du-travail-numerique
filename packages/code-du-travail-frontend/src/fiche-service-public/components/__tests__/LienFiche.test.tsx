import React from "react";
import { render } from "@testing-library/react";
import dataMock from "./mocks/lienFiche.json";
import { ElementBuilder } from "../ElementBuilder";

describe("LienFiche", () => {
  it("should display link", () => {
    const { container } = render(<ElementBuilder data={dataMock} />);
    expect(container).toMatchSnapshot();
  });
});
