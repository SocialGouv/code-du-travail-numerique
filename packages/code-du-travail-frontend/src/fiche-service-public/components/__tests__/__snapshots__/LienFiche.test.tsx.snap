// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LienFiche should display link 1`] = `
<div>
  <p>
    <a
      href="https://www.service-public.fr/particuliers/vosdroits/F24411"
    >
      Départ négocié (difficultés économiques)
    </a>
  </p>
</div>
`;
